ddtrace/contrib/urllib3/patch.py

```diff
@@ -0,0 +1,166 @@
+import urllib3
+from six.moves.urllib.parse import urlparse, urlunparse
+
+import ddtrace
+from ddtrace import config
+from ddtrace.vendor.wrapt import wrap_function_wrapper as _w
+from ddtrace.http import store_request_headers, store_response_headers
+
+from ...ext import http
+from ...utils.http import sanitize_url_for_tag
+from ...utils.wrappers import unwrap as _u
+from ...utils.formats import asbool, get_env
+from ...propagation.http import HTTPPropagator
+from ...constants import ANALYTICS_SAMPLE_RATE_KEY
+
+
+# Ports which, if set, will not be used in hostnames/service names
+DROP_PORTS = (80, 443)
```

Comment:
If a user of `urllib3` passes the default port for `HTTPConnectionPool` or `HTTPSConnectionPool`, we would be stripping the port from the url. That seems reasonable though an additional step in our code that might not be needed.

I'm more confused by the situation where these default ports are being set without them being specified by the user. Let's make sure we understand why that's the case. Perhaps its specific to certain versions of the library?

At least with 1.25.7, I only see this happen when using `connection_from_url`:

```
>>> from urllib3 import HTTPConnectionPool
>>> import urllib3
>>> p1 = HTTPConnectionPool('ajax.googleapis.com')
>>> p1.scheme, p1.host, p1.port
('http', 'ajax.googleapis.com', None)
>>> p2 = urllib3.connectionpool.connection_from_url("http://ajax.googleapis.com")
>>> p2.scheme, p2.host, p2.port
('http', 'ajax.googleapis.com', 80)
>>> p3 = urllib3.connectionpool.connection_from_url("https://ajax.googleapis.com")
>>> p3.scheme, p3.host, p3.port
('https', 'ajax.googleapis.com', 443)
>>> from urllib3 import HTTPSConnectionPool
>>> p4 = HTTPSConnectionPool('ajax.googleapis.com')
>>> p4.scheme, p4.host, p4.port
('https', 'ajax.googleapis.com', None)
```

Reply:
One of the common ways of making requests via urllib3 is by using `urllib3.PoolManager().request(...)`, which abstracts the underlying connection pools from the user. As part of this abstraction, ports are assigned to the selected ConnectionPool class as a feature of the library:

https://github.com/urllib3/urllib3/blob/master/src/urllib3/poolmanager.py#L217-L221

This inferred port is finally passed to the constructor here: https://github.com/urllib3/urllib3/blob/master/src/urllib3/poolmanager.py#L268

This happens whether a port is explicitly declared in the request URL or not. 

```
>>> from ddtrace import patch
>>> patch(urllib3=True)
>>> import urllib3
>>> http = urllib3.PoolManager()
>>> r = http.request('GET', 'http://httpbin.org/robots.txt')
# In the patched function:
(Pdb) obj.port
80
```

We can monkey-patch at a higher level, detect if the port is set on the url, and set a flag on the connection pool class. But let's see if customers actually request that feature before going that route.

Reply:
Thanks for the explanation! 👍